server/market/orderrouter.go

```diff
@@ -518,10 +398,112 @@ func (r *OrderRouter) handleMarket(user account.AccountID, msg *msgjson.Message)
 		msgID: msg.ID,
 	}
 
+	return r.processTrade(oRecord, tunnel, assets, market.Coins, sell, 0, market.RedeemSig, market.Serialize())
+}
+
+// processTrade checks that the trade is valid and submits it to the market.
+func (r *OrderRouter) processTrade(oRecord *orderRecord, tunnel MarketTunnel, assets *assetSet,
+	coins []*msgjson.Coin, sell bool, rate uint64, redeemSig *msgjson.RedeemSig, sigMsg []byte) *msgjson.Error {
+
+	fundingAsset := assets.funding
+	user := oRecord.order.User()
+	trade := oRecord.order.Trade()
+
+	// If the receiving asset is account-based, we need to check that they can
+	// cover fees for the redemption, since they can't be subtracted from the
+	// received amount.
+	receivingBalancer, isToAccount := assets.receiving.Backend.(asset.AccountBalancer)
+	if isToAccount {
+		if redeemSig == nil {
+			log.Info("user %s did not include a RedeemSig for received asset %s", user, assets.receiving.Symbol)
+			return msgjson.NewError(msgjson.OrderParameterError, "no redeem address verification included for asset %s", assets.receiving.Symbol)
+		}
+
+		acctAddr := trade.ToAccount()
+		if err := receivingBalancer.ValidateSignature(acctAddr, redeemSig.PubKey, sigMsg, redeemSig.Sig); err != nil {
+			log.Info("user %s failed redeem signature validation for order %s: %v",
+				user, oRecord.order.ID(), err)
+			return msgjson.NewError(msgjson.SignatureError, "redeem signature validation failed")
+		}
+
+		if !r.sufficientAccountBalance(acctAddr, oRecord.order, &assets.receiving.Asset, tunnel) {
+			return msgjson.NewError(msgjson.FundingError, "insufficient balance")
+		}
+	}
+
+	// If the funding asset is account-based, we'll check balance and submit the
+	// order immediately, since we don't need to find coins.
+	fundingBalancer, isAccountFunded := assets.funding.Backend.(asset.AccountBalancer)
+	if isAccountFunded {
+		// Validate that the coins are correct for an account-based-asset-funded
+		// order. There should be 1 coin, 1 sig, 1 pubkey, and no redeem script.
+		if len(coins) != 1 {
+			log.Info("user %s submitted an %s-funded order with %d coin IDs", user, assets.funding.Symbol, len(coins))
+			return msgjson.NewError(msgjson.OrderParameterError, "account-type asset funding requires exactly one coin ID")
+		}
+		acctProof := coins[0]
+		if len(acctProof.PubKeys) != 1 || len(acctProof.Sigs) != 1 || len(acctProof.Redeem) > 0 {
```

Comment:
Working through things here:

- The "coin ID" in both `acctProof.ID` and `trade.Coins[0]` are equal, as evident in the callers, `handleMarket` and `handleLimit` who populate the `order.Trade.Coins` from each `msgjson.{Market,Limit}Order.Trade.Coins[i].ID`
- These coin ID encodings deviate from the new transaction id/hash convention for coin IDs being established in https://github.com/decred/dcrdex/pull/1307.  Hence `ValidateCoinID` cannot be used on them.
- They are also the same as `order.(*Trade).FromAccount` before hex string encoding.
- `order.(*Trade).ToAccount` is the same as `order.Trade.Address` (`order.(*Trade).SwapAddress`)

Mostly I want to be sure this is all the intended design and I'm not missing anything.

Reply:
That mostly seems right. The term "coin ID" is so general now as to be confusing. The definition of coin ID need to be evaluated separately for each message and asset. 

`ValidateCoinID` is in a weird place right now, I guess. It's used for funding/swap/redeem coins for utxo-based assets, but only for swap/redeem coins for acct-based. I wouldn't mind breaking that method into as many as 3 separate methods, `ValidateSwapCoin`, `ValidateRedeemCoin`, and `ValidateFundingCoin` (`OutputTracker`-only). We could also combine the swap and redeem methods into something like `ValidateTradeCoin` for now. Or we could just leave it. It's all terrible.

Reply:
Let's solve with docs for now rather than end up with a ton of new methods and wrappers and stuff